### PR TITLE
Fix updating filters in the filter header for multi-stage queries

### LINKED ISSUE #24994: Can't change filters via 'Show filters'
**Describe the bug**

In a saved question, when I clicked on "Show filters" and try to change any of the filters (date, is empty, etc), the Update filter button grays out immediately. 

**Logs**
None

**To Reproduce**
1. Go to any saved question with filters
2. Click on "Show filters"
3. Try to change any filter's value (is empty to not empty, date to another date, etc)
4. Update filter button grays out immediately

**Expected behavior**
Update filter button shouldn't gray out, so the new filter can be applied. 

**Screenshots**

n/a as it's pretty simple / screenshot won't provide much value

**Information about your Metabase Installation:**

You can get this information by going to Admin -> Troubleshooting.

- Your browser and the version:  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36",

- Your operating system: MacIntel
- Your databases: Redshift
- Metabase version: master / 1fca705
- Metabase hosting environment: Metabase Cloud
- Metabase internal database: Postgres

**Severity**
How severe an issue is this bug to you? Is this annoying, blocking some users, blocking an upgrade or blocking your usage of Metabase entirely?
It's a bit inconvenient as I need to go to the question's editor page to make changes instead. 

**Additional context**
None

### PR DESCRIPTION
**Pull request by ranquild**
Fixes https://github.com/metabase/metabase/issues/24994
Reproduces https://github.com/metabase/metabase/issues/24994
Regression since https://github.com/metabase/metabase/pull/24729

How to test:
- Follow the steps from this comment https://github.com/metabase/metabase/issues/24994#issuecomment-1227208189
- Please note that there is a separate issue https://github.com/metabase/metabase/issues/11957 since 0.32 that is not fixed by this PR

### CODE REVIEW

**Comment by ranquild** [comment]
@nemanjaglumac I've closed https://github.com/metabase/metabase/pull/25065 in favour of this PR. Here I found the exact PR in 0.44 that caused the issue and logically reverted that part.

File: frontend/src/metabase/query_builder/components/filters/FilterPopover/FilterPopover.tsx [hunk 1]
```diff
@@ -146,8 +146,7 @@ export default function FilterPopover({
   };
 
   const handleFilterChange = (mbql: any[] = []) => {
-    const newFilter = new Filter(mbql, filter?.index(), query);
-
+    const newFilter = filter ? filter.set(mbql) : new Filter(mbql, null, query);
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/filters/FilterPopover/FilterPopover.tsx:149]: Here how it looked like in 0.43 https://github.com/metabase/metabase/blob/c5aebf70dc164bde6d2247c0bf1762a4203f9d9a/frontend/src/metabase/query_builder/components/filters/FilterPopover.tsx#L141

Comment (ranquild) [reply frontend/src/metabase/query_builder/components/filters/FilterPopover/FilterPopover.tsx:149]: The root cause is that `filter.set` uses the query for the same stage level, while `query` here is the top-level query.

File: frontend/src/metabase/query_builder/components/filters/FilterPopover/FilterPopover.tsx [hunk 2]
```diff
@@ -84,7 +84,7 @@ export default function FilterPopover({
 
   // if the underlying query changes (e.x. additional metadata is loaded) update the filter's query
   useEffect(() => {
-    if (filter && previousQuery && query !== previousQuery) {
+    if (filter && filter.query() === previousQuery && query !== previousQuery) {
```
Comment (gusaiani) [inline frontend/src/metabase/query_builder/components/filters/FilterPopover/FilterPopover.tsx:87]: `filter?.query()`

**Review by gusaiani — APPROVED** [review]
👍, one very minor comment inlined.

**Pull request merged**